deepchem/feat/base_classes.py

```diff
@@ -6,16 +6,64 @@
 import numpy as np
 import multiprocessing
 
-__author__ = "Steven Kearnes"
-__copyright__ = "Copyright 2014, Stanford University"
-__license__ = "BSD 3-clause"
+logger = logging.getLogger(__name__)
 
 
 def _featurize_complex(featurizer, mol_pdb_file, protein_pdb_file, log_message):
   logging.info(log_message)
   return featurizer._featurize_complex(mol_pdb_file, protein_pdb_file)
 
 
+class Featurizer(object):
+  """Abstract class for calculating a set of features for a datapoint.
+
+  This class is abstract and cannot be invoked directly. You'll
+  likely only interact with this class if you're a developer. In
+  that case, you might want to make a child class which
+  implements the `_featurize` method for calculating features for
+  a single datapoints if you'd like to make a featurizer for a
+  new datatype.
+  """
+
+  def featurize(self, datapoints, log_every_n=1000):
+    """Calculate features for datapoints.
+
+    Parameters
+    ----------
+    datapoints: iterable 
+       A sequence of objects that you'd like to featurize. Subclassses of
+       `Featurizer` should instantiate the `_featurize` method that featurizes
+       objects in the sequence.
+
+    Returns
+    -------
+    A numpy array containing a featurized representation of `datapoints`.
+    """
+    datapoints = list(datapoints)
+    features = []
+    for i, point in enumerate(datapoints):
+      if i % log_every_n == 0:
+        logger.info("Featurizing datapoint %i" % i)
+      if point is not None:
+        features.append(self._featurize(point))
+      else:
+        features.append(np.array([]))
```

Comment:
Maybe featurizing sometimes failed. It is better to clarify the error like this.

```python
try:
  features.append(self._featurize(point))
except:
  logger.info("Failed to featurize datapoint %i. This was converted to an empty array." % i)
  features.append(np.array([]))
```

Reply:
Good suggestion! Added in